File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        # 앞에서부터 최고점, 현재 지점 높이를 가지고 가다가
        # 현재의 최고점과 같거나 보다 높은 기둥을 만나면 아래 두가지를 수행
        # 1. 최고점 갱신
        # 2. 현재까지 저장된 물 높이를 총 물 량에 갱신
        whole_water = 0
        cur_water = 0
        peak = 0
        
        for h in height:
            if h >= peak:
                peak = h
                whole_water += cur_water
                cur_water = 0
            else:
                cur_water += peak - h
        
        # 최고점이 나온 후 끝까지는 계산되지 않는다.
        # 따라서 최고점이 나올 때까지 거꾸로 한 번 수행
        cur_water = 0
        end = peak
        peak = 0
        
        height.reverse()
        for h in height:
            if h == end:
                whole_water += cur_water
                break
                
            if h >= peak:
                peak = h
                whole_water += cur_water
                cur_water = 0
            else:
                cur_water += peak - h
        
        return whole_water

File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int answer = 0;
        
        // 최대 값이 존재하는 index를 찾는다.
        int maxIdx = 0;
        int maxHeight = -1;
        for(int i = 0, size = height.length; i < size; ++i) {
            if (height[i] > maxHeight) {
                maxHeight = height[i];
                maxIdx = i;
            }
        }
        
        // 처음 ~ 첫 번째 최대 높이 기둥이 존재하는 곳 까지 빗물의 양
        int curMaxHeight = 0;
        int sum = 0;
        for (int i = 0; i <= maxIdx; ++i) {
            // 이전 최대 높이 이상의 기둥을 만난다면
            if (curMaxHeight < height[i]) {
                curMaxHeight = height[i];
                answer += sum;
                sum = 0;
            } else {
                sum += curMaxHeight - height[i];
            }
        }
        
        // 첫 번째 최대 높이 기둥 ~ 끝까지의 빗물의 양
        // 뒤에서부터 거꾸로 오며 위의 로직을 반복한다.
        curMaxHeight = 0;
        for (int i = height.length - 1; i >= maxIdx; --i) {
            if (curMaxHeight <= height[i]) {
                curMaxHeight = height[i];
                answer += sum;
                sum = 0;
            } else {
                sum += curMaxHeight - height[i];
            }
        }
        
        return answer;
    }
}